### PR TITLE
Allow AuthorizationPolicys to reference ServiceAccounts

### LINKED ISSUE #8565: rfc: Allow `AuthorizationPolicy` to target `ServiceAccount` directly?
Currently, we require that `AuthorizationPolicy` resources reference a `MeshTLSAuthentication` resource to authorize a pod by its service account. Like so:

```yaml
apiVersion: policy.linkerd.io/v1alpha1
kind: AuthorizationPolicy
metadata:
  name: linkerd-admin-metrics-viz
  namespace: emojivoto
spec:
    targetRef:
      ...
    requiredAuthenticationRefs:
      - kind: MeshTLSAuthentication
        name: linkerd-viz
---
apiVersion: policy.linkerd.io/v1alpha1
kind: MeshTLSAuthentication
metadata:
  name: linkerd-viz
  namespace: emojivoto
spec:
    identityRefs:
      - kind: ServiceAccount
        name: prometheus
        namespace: linkerd-viz
```

I think we can simplify this in many cases if we support referencing the `ServiceAccount` directly:

```yaml
apiVersion: policy.linkerd.io/v1alpha1
kind: AuthorizationPolicy
metadata:
  name: linkerd-admin-metrics-viz
  namespace: emojivoto
spec:
    targetRef:
      ...
    requiredAuthenticationRefs:
      - kind: ServiceAccount
        name: prometheus
        namespace: linkerd-viz
```

This is a more general assertion--we're requiring that the client _somehow_ used a service account to authenticate to the server. Whether that was using mesh TLS or not is an implementation detail (that could be changed in the future).

This feels like a nice quality-of-life improvement.

Thoughts?

### PR DESCRIPTION
**Pull request by kleimkuhler**
Closes #8565.

With this change, AuthorizationPolicys can now reference ServiceAccounts for their target authentications. This allows users to avoid the requirement of creating  a MeshTLSAuthentication resource that references a single ServiceAccount.

The policy admission controller only allows an AuthorizationPolicy to reference a single MeshTLSAuthentication _or_ a ServiceAccount; it cannot reference both. Additionally, if a ServiceAccount is reference it can onl be a single one — similar to MeshTLSAuthentications.

Signed-off-by: Kevin Leimkuhler <kleimkuhler@icloud.com>

### CODE REVIEW

File: policy-controller/k8s/index/src/index.rs [hunk 1]
```diff
@@ -1120,6 +1120,22 @@ impl PolicyIndex {
                 );
             }
         }
+        for tgt in spec.authentications.iter() {
```
Comment (adleong) [inline policy-controller/k8s/index/src/index.rs:1123]: Can we combine these two loops and process both MeshTLS and ServiceAccount in the same loop?

```
for tgt in spec.authentications.iter() {
  match tgt {
    AuthenticationTarget::MeshTLS { ref namespace, ref name } => { ... },
    AuthenticationTarget::ServiceAccount { ref namespace, ref name } => { ... },
    AuthenticationTarget::Network { ref namespace, ref name } => { ... },
  }
}
```

**Review by mateiidavid — APPROVED** [review]
Looks good to me, modulo Alex's suggestion.

File: policy-controller/k8s/index/src/index.rs [hunk 2]
```diff
@@ -1092,65 +1092,74 @@ impl PolicyIndex {
...
-                            bail!("policy must not include multiple MeshTLSAuthentications");
+            match tgt {
+                AuthenticationTarget::Network {
+                    ref namespace,
+                    ref name,
+                } => {
+                    let namespace = namespace.as_deref().unwrap_or(&self.namespace);
+                    tracing::trace!(ns = %namespace, %name, "Finding NetworkAuthentication");
+                    if let Some(ns) = all_authentications.by_ns.get(namespace) {
+                        if let Some(authn) = ns.network.get(name).as_ref() {
```
Comment (adleong) [inline policy-controller/k8s/index/src/index.rs:1105]: I think this `.as_ref()` is unnecessary and results in a double ref

**Review by adleong — COMMENTED** [review]
Looks good.

TIOLI, but I find the continues nested inside if blocks make it a bit hard to reason about the control flow.  This is just personal style, but what do you think of a slightly more functional style?

e.g.

```
diff --git a/policy-controller/k8s/index/src/index.rs b/policy-controller/k8s/index/src/index.rs
index 7723d0a93..a924fe645 100644
--- a/policy-controller/k8s/index/src/index.rs
+++ b/policy-controller/k8s/index/src/index.rs
@@ -11,7 +11,7 @@ use crate::{
     pod, server, server_authorization, ClusterInfo,
 };
 use ahash::{AHashMap as HashMap, AHashSet as HashSet};
-use anyhow::{bail, Result};
+use anyhow::{anyhow, bail, Result};
 use linkerd_policy_controller_core::{
     AuthorizationRef, ClientAuthentication, ClientAuthorization, IdentityMatch, InboundServer,
     IpNet, Ipv4Net, Ipv6Net, NetworkMatch, ProxyProtocol, ServerRef,
@@ -1101,24 +1101,25 @@ impl PolicyIndex {
                 } => {
                     let namespace = namespace.as_deref().unwrap_or(&self.namespace);
                     tracing::trace!(ns = %namespace, %name, "Finding NetworkAuthentication");
-                    if let Some(ns) = all_authentications.by_ns.get(namespace) {
-                        if let Some(authn) = ns.network.get(name).as_ref() {
-                            tracing::trace!(ns = %namespace, %name, nets = ?authn.matches, "Found NetworkAuthentication");
-                            // There can only be a single required NetworkAuthentication. This is
-                            // enforced by the admission controller.
-                            if networks.is_some() {
-                                bail!("policy must not include multiple NetworkAuthentications");
-                            }
-                            let nets = authn.matches.clone();
-                            networks = Some(nets);
-                            continue;
-                        }
+                    let authn = all_authentications
+                        .by_ns
+                        .get(namespace)
+                        .and_then(|ns| ns.network.get(name))
+                        .ok_or_else(|| {
+                            anyhow!(
+                                "could not find NetworkAuthentication {} in namespace {}",
+                                name,
+                                namespace
+                            )
+                        })?;
+                    tracing::trace!(ns = %namespace, %name, nets = ?authn.matches, "Found NetworkAuthentication");
+                    // There can only be a single required NetworkAuthentication. This is
+                    // enforced by the admission controller.
+                    if networks.is_some() {
+                        bail!("policy must not include multiple NetworkAuthentications");
                     }
-                    bail!(
-                        "could not find NetworkAuthentication {} in namespace {}",
-                        name,
-                        namespace
-                    );
+                    let nets = authn.matches.clone();
+                    networks = Some(nets);
                 }
                 AuthenticationTarget::MeshTLS {
                     ref namespace,
@@ -1126,24 +1127,25 @@ impl PolicyIndex {
                 } => {
                     let namespace = namespace.as_deref().unwrap_or(&self.namespace);
                     tracing::trace!(ns = %namespace, %name, "Finding MeshTLSAuthentication");
-                    if let Some(ns) = all_authentications.by_ns.get(namespace) {
-                        if let Some(authn) = ns.meshtls.get(name) {
-                            tracing::trace!(ns = %namespace, %name, ids = ?authn.matches, "Found MeshTLSAuthentication");
-                            // There can only be a single required MeshTLSAuthentication. This is
-                            // enforced by the admission controller.
-                            if identities.is_some() {
-                                bail!("policy must not include multiple MeshTLSAuthentications");
-                            }
-                            let ids = authn.matches.clone();
-                            identities = Some(ids);
-                            continue;
-                        }
+                    let authn = all_authentications
+                        .by_ns
+                        .get(namespace)
+                        .and_then(|ns| ns.meshtls.get(name))
+                        .ok_or_else(|| {
+                            anyhow!(
+                                "could not find MeshTLSAuthentication {} in namespace {}",
+                                name,
+                                namespace
+                            )
+                        })?;
+                    tracing::trace!(ns = %namespace, %name, ids = ?authn.matches, "Found MeshTLSAuthentication");
+                    // There can only be a single required MeshTLSAuthentication. This is
+                    // enforced by the admission controller.
+                    if identities.is_some() {
+                        bail!("policy must not include multiple MeshTLSAuthentications");
                     }
-                    bail!(
-                        "could not find MeshTLSAuthentication {} in namespace {}",
-                        name,
-                        namespace
-                    );
+                    let ids = authn.matches.clone();
+                    identities = Some(ids);
                 }
                 AuthenticationTarget::ServiceAccount {
                     ref namespace,
```

File: policy-controller/k8s/index/src/index.rs [hunk 3]
```diff
@@ -1092,65 +1092,74 @@ impl PolicyIndex {
...
+                            tracing::trace!(ns = %namespace, %name, nets = ?authn.matches, "Found NetworkAuthentication");
```
Comment (hawkw) [inline policy-controller/k8s/index/src/index.rs:1106]: maybe the `namespace` and `name` fields should be added to a span?
```suggestion
                    let _span = tracing::trace_span!("network", ns = %namespace, %name).entered();
                    tracing::trace!("Finding NetworkAuthentication...");
                    if let Some(ns) = all_authentications.by_ns.get(namespace) {
                        if let Some(authn) = ns.network.get(name).as_ref() {
                            tracing::trace!(nets = ?authn.matches, "Found NetworkAuthentication");
```

File: policy-controller/k8s/index/src/index.rs [hunk 4]
```diff
@@ -1092,65 +1092,74 @@ impl PolicyIndex {
...
-                            bail!("policy must not include multiple NetworkAuthentications");
+                AuthenticationTarget::MeshTLS {
+                    ref namespace,
+                    ref name,
+                } => {
+                    let namespace = namespace.as_deref().unwrap_or(&self.namespace);
+                    tracing::trace!(ns = %namespace, %name, "Finding MeshTLSAuthentication");
+                    if let Some(ns) = all_authentications.by_ns.get(namespace) {
+                        if let Some(authn) = ns.meshtls.get(name) {
+                            tracing::trace!(ns = %namespace, %name, ids = ?authn.matches, "Found MeshTLSAuthentication");
```
Comment (hawkw) [inline policy-controller/k8s/index/src/index.rs:1131]: similarly, maybe
```suggestion
                    let namespace = namespace.as_deref().unwrap_or(&self.namespace);
                    let _span = tracing::trace_span!("mesh_tls", ns = %namespace, %name).entered();
                    tracing::trace!("Finding MeshTLSAuthentication...");
                    if let Some(ns) = all_authentications.by_ns.get(namespace) {
                        if let Some(authn) = ns.meshtls.get(name) {
                            tracing::trace!(ids = ?authn.matches, "Found MeshTLSAuthentication");
```

File: policy-controller/src/admission.rs [hunk 5]
```diff
@@ -227,13 +236,20 @@ impl Validate<AuthorizationPolicySpec> for Admission {
             bail!("only a single NetworkAuthentication may be set");
         }
 
-        if mtls_authns_count + net_authns_count < spec.required_authentication_refs.len() {
+        if mtls_authns_count + sa_authns_count > 1 {
+            bail!("only MeshTLSAuthentication or ServiceAccount may be set");
+        }
+
```
Comment (olix0r) [inline policy-controller/src/admission.rs:242]: We can do this check before we do any of the net checks (i.e. move this up to 229).

File: policy-controller/src/admission.rs [hunk 6]
```diff
@@ -227,13 +236,20 @@ impl Validate<AuthorizationPolicySpec> for Admission {
             bail!("only a single NetworkAuthentication may be set");
         }
 
-        if mtls_authns_count + net_authns_count < spec.required_authentication_refs.len() {
+        if mtls_authns_count + sa_authns_count > 1 {
+            bail!("only MeshTLSAuthentication or ServiceAccount may be set");
```
Comment (olix0r) [inline policy-controller/src/admission.rs:240]: This error message is not clear to me... It seems to insinuate that a NetworkAuthentication can't be set.

```suggestion
            bail!("a MeshTLSAuthentication and ServiceAccount may not be set together");
```

**Review by olix0r — CHANGES_REQUESTED** [review]
It's minor, but I'd prefer not to join the network & identity loops in this change -- the diff would be smaller without that change and I don't think it particularly helps readability.

**Comment by kleimkuhler** [comment]
Networks are processed in a separate loop and we ignore them in the match for identities (`_network => {}`).

File: policy-controller/k8s/index/src/index.rs [hunk 7]
```diff
@@ -1093,31 +1093,47 @@ impl PolicyIndex {
...
+                                namespace
+                            )
+                        })?;
+                    tracing::trace!(ids = ?authn.matches, "Found MeshTLSAuthentication");
+                    if identities.is_some() {
+                        bail!("policy must not include multiple MeshTLSAuthentications");
                     }
+                    let ids = authn.matches.clone();
+                    identities = Some(ids);
+                    continue;
```
Comment (adleong) [inline policy-controller/k8s/index/src/index.rs:1121]: I think this `continue` is redundant

**Review by adleong — APPROVED** [review]

**Review by olix0r — APPROVED** [review]

**Pull request merged**